### ISSUE TITLE
Error using pthread: Pthread 0x0 completed its main entry point with an `unwind`, keeping the worker alive for asynchronous operation.

### ISSUE THREAD

**Issue by 0-0alephant** [emscripten-core/emscripten#19074]
Please include the following in your bug report:

**Version of emscripten/emsdk:**
Please include the output `emcc -v` here
```
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.32 (eab98adf462c39f3c31d348331c4830bcaa36949)
clang version 17.0.0 (https://github.com/llvm/llvm-project df82394e7a2d06506718cafa347bf7827c79fc4f)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: D:\emsdk\upstream\bin
```
**Failing command line in full:**
If this is compile or link-time failure please include the full failing command
along with its entire output.
```module.worker.js:1 Pthread 0x0 completed its main entry point with an `unwind`, keeping the worker alive for asynchronous operation.```

**Full link command and output with `-v` appended:**
Even for runtime issues it helps a lot if you can include the full link command.
Adding `-v` to the link command will show all of the sub-commands run which
can help us diagnose your issue.
```
	 em++ -O3 \
                -std=c++17 \
		-s ASSERTIONS=1 \
		-s ALLOW_MEMORY_GROWTH=1 \
		-Wno-experimental \
	 	-s MAIN_MODULE \
		 -s MODULARIZE \
		 -s EXPORT_ES6 \
		 -s ENVIRONMENT='web,worker'  \
		 -s EXPORT_NAME=newModule \
		 -sFULL_ES3 \
		 -sASYNCIFY \
		-pthread \
		-sPTHREAD_POOL_SIZE=navigator.hardwareConcurrency \
		 -s USE_GLFW=3 \
		 -lwebsocket.js \
		 -s WASM=1 \
                main.cc\
               ... some omitted ...
                -o module.mjs

```

I think the problem mainly occured because of I enabled ASYNCIFY, 

this is testing code , without this part, the code works fine.

```
// MainModule  
#include <pthread.h>
void* fn(void* ){
    printf("%d\n", 1);
    pthread_exit(0);
}

int main(){


  pthread_t bg_thread;
  if(pthread_create(&bg_thread, NULL, fn, NULL)){
    printf("error: pthread create failed\n");
    return 1;
  }

  if(pthread_join(bg_thread, NULL)){
    printf("join failed\n");
  }
}
```

if I comment out ``` pthread_exit(0);``` it will give me a different error:

![Screenshot 2023-03-26 195128](https://user-images.githubusercontent.com/128656803/227773779-e8248ddf-14a5-409d-9a27-e83cd8b500c3.jpg)

if I put  ``` pthread_exit(0);```  back on, the error becomes:

![Screenshot 2023-03-26 195244](https://user-images.githubusercontent.com/128656803/227773827-fef776c5-be69-4a40-8990-5b7254f57842.jpg)

by the way, I set ```NoRuntimeExit``` to true, since I need  the runtime still present after main.

more info if I click the call trace:

```

runtime error: unreachable:
 at module.wasm:
      global.get $env.__asyncify_state
      i32.eqz
      if
        unreachable
      end
      unreachable
    end $label0
    local.set $var2
```

**Comment by sbc100** [comment]
The `Pthread 0x0 completed its main entry point with an unwind` error should be fixed with #19048.

Regarding the second issue related to asyncify and pthreads, perhaps you can open a seperate issue for that? 

Regarding the setting of `NoRuntimeExit`, where are you doing that?  i.e. on which thread(s)?    A better way to set that variable would be with `emscripten_exit_with_live_runtime()` or `emscripten_runtime_keepalive_push()`.

**Comment by 0-0alephant** [comment]
[#19048](https://github.com/emscripten-core/emscripten/pull/19048) works .  I'll post another issue if there's problem when using asyncify, ```NoRuntimeExit``` mainly used since I want to call some function in C side later on when some button is clicked , I've changed ```NoRuntimeExit ```to ```emscripten_exit_with_live_runtime()```,  no error messages so far, it seems nice.
